icons/src/fullcolor/render-bitmaps-verbose.py

```diff
@@ -1,234 +1,217 @@
-                            inkscape_render_rect(self.path, id, dpi, outfile)
-                            if args.verbose:
-                                print("├─ Rendered new \"" + outfile + "\"")
-                            new_renders += 1
 
-                        # If PNG exists, compare modify time to that of SVG
-                        else:
-                            stat_in = os.stat(self.path)
-                            stat_out = os.stat(outfile)
-
-                            # If SVG is newer than PNG, replace PNG with updated version
-                            if stat_in.st_mtime > stat_out.st_mtime:
-                                inkscape_render_rect(self.path, id, dpi, outfile)
-                                if args.verbose:
-                                    print("├─ Rendered updated \"" + outfile + "\"")
-                                # print("Rendered updated " + outfile)
-                                updated_renders += 1
-
-                            # If PNG is newer than SVG, leave PNG as is
-                            else:
-                                if args.verbose:
-                                    print("├─ \"" + outfile + "\" is newer than SVG")
-                                skipped_renders += 1
-                if args.verbose:
-                    print("├────────────────────────────────┤")
-                if args.svg is None:
-                    print("")
-                    print("┌────────────────────────────────┐")
-                    print("│ Directory: " + self.context)
-                    print("│ Icon Name: " + self.icon_name)
-                    print("├────────────────────────────────┤")
-                    print("├─ Rendered %d new PNGs" % new_renders)
-                    print("├─ Rendered %d updated PNGs" % updated_renders)
-                    print("├─ Skipped %d up-to-date PNGs" % skipped_renders)
-                    print("└────────────────────────────────┘")
-                    print("")
-
-        def characters(self, chars):
-            self.chars += chars.strip()
+def render_category(args, category_to_render):
 
     # If invocation includes a file name, try to process it
-    if args.svg:
-        file = os.path.join(SRC, args.svg + '.svg')
+    if args.individual_icons:
+        file = os.path.join(category_to_render, args.individual_icons + '.svg')
 
         if os.path.exists(os.path.join(file)):
             print("├─ Rendering from \"" + os.path.join(file) + "\"")
-            handler = ContentHandler(file, True, filter=args.filter)
+            #
+            # ContentHandler() implements an interface for xml.sax.parse()
+            #
+            handler = ContentHandler(file, args, True)
+            #
+            # xml.sax.parse() uses the various ContentHandler() methods behind the scenes
+            #
             xml.sax.parse(open(file), handler)
             return True
         else:
```

Comment:
there is no need for an `else` branch after `return`

Reply:
Looking at the code, the `else` branch happens if the user specifies the name of an icon name that doesn’t exist. Or something. Again, basically all the comments and the `print()` statements are from when I was debugging the initial version of the script, so there are a fair number of extra failsafes built in. I should probably try running this through `pylint`.

Reply:
Well, it looks like the `—categories` parser is broken so I wasn’t able to reach that `else` branch quite yet:
```bash
$ ./render-bitmaps-verbose.py --categories apps --individual_icons thisappdoesnotexist
┌────────────────────────────────┐
│ Rendering from command-line argument "thisappdoesnotexist"
├────────────────────────────────┤
Source path "./a" does not exist.
Source path "./p" does not exist.
Source path "./p" does not exist.
Source path "./s" does not exist.
└────────────────────────────────┘

Failed to process "thisappdoesnotexist" in .
```

Reply:
Considering the output above, even if good looking, the logs can be formatted without those borders. What do you think?

Reply:
You mean the ASCII “box” outlines? I specifically found they helped with visual parsing extremely long sections of output. Like, CMake has weird specific formatting, too, for the purpose of visually parsing the (extremely verbose) output.

The issue in the above is that it’s parsing “apps” as “a” “p” “p” “s”.

Reply:
Yes, I agree they help a little, but as well it is easy to break like the following
```
┌────────────────────────────────┐
│ Rendering from command-line argument "thisappdoesnotexist"
├────────────────────────────────┤
```

which is not really nice to see.

I mean, there are python modules to handle such prints, but it seems overkill to me. A better option would be to move to a proper logger, but it is better to move it to a different PR

Reply:
> The issue in the above is that it’s parsing “apps” as “a” “p” “p” “s”.

sorry if you already detected it, but this is because `args.category` is defined as single str value

```python
parser.add_argument(
        '--categories',
        type=str,
        nargs='?',
```

`nargs` should be `*` to get a list of zero or more values